### ISSUE TITLE
[RFE] `fail2ban-client reban` to restore bans after flushing of iptables

### ISSUE THREAD

**Issue by LeonardoGentile** [fail2ban/fail2ban#1680]
Hello, this is more a question than an issue, I couldn't find an answer anywhere else. 
On system startup (debian/ubuntu) I have a script that restore the iptables rules `/etc/network/if-pre-up.d/load-iptables-rules` from a file. After this is executed the system automatically start the fail2ban service and append its rules and the previously banned ips to iptables rules. So far so good.

The problem come when something (or I manually) trigger an `iptables-restore` or a manual `iptables -F`. In this case of course the failban rules and banned ips are not restored. In order for this to work I have to manually `service fail2ban restart`.

So my question: is there any way for fail2ban to listen to these kind of events (iptables-restore, iptables -F, in general anything external wiping out the fail2ban iptables rules) and possibly trigger a reload of fail2ban in order to append again the f2b rules to iptables?

**Comment by sebres** [comment]
> is there any way for fail2ban to listen to these kind of events

No. It would be at least very strange if fail2ban would do this, because it was not written for such purposes and fail2ban can much more actions as iptables.
Truth be told, I know no service, that would do something similar.

Solutions:
- write better script as `iptables-restore` that will not touch entries generated from fail2ban (this would be the right way);
- or write another script that will filter chains starting with "f2b-" to separate dump and restore it additionally after `iptables-restore`;
- if these "something" (that trigger iptables -F) is a service, then add dependencies to fail2ban, so it guarantees automatically restart of fail2ban, after restarting of this service;
- rewrite or mock-up action `iptables-*` with your own, that will silently ignore such kinds of "errors" (but then you'll not get all the IPs banned back, after applying of iptables-restore);

**Comment by LeonardoGentile** [comment]
Thank you very much for your time 👍 

I'm trying with the implementation of the first point, for example.
Let's say I have a set of fixed rules in `/etc/my.fixed.rules` that on system startup and on ifup/ifdown will be loaded with `iptable-restore`. 

Now, just before restoring I could do the following, make an `iptables-save` to `/tmp/f2b-rules`, then with a script I inspect all the rules and if there is any (at all) that doesn't have the word `fail2ban` in the rule/chain then it will be removed. So now I have two files:

- /etc/my.fixed.rules
- /tmp/f2b-rules (cleaned up)

I can do first a restore of my fixed rules with `iptables-restore < /etc/my.fixed.rules` then a restore of the fail2ban rules `iptables-restore -n < /tmp/f2ban-rules`. `-n` is important not to flush the previously imported fixed rules.

So far so good, the only concerning thing is the order or the rules. When f2b add the rules automatically when starting the service the order is something like:

```
Chain INPUT (policy ACCEPT)
target     prot opt source               destination
# F2B Rules
fail2ban-php-url-fopen  tcp  --  anywhere             anywhere             multiport dports http,https
fail2ban-apache-nohome  tcp  --  anywhere             anywhere             multiport dports http,https
# My Fixed Rules
ACCEPT     all  --  anywhere             anywhere
REJECT     all  --  anywhere             loopback/8           reject-with icmp-port-unreachable
ACCEPT     all  --  anywhere             anywhere             state RELATED,ESTABLISHED
ACCEPT     tcp  --  anywhere             anywhere             tcp dpt:http
ACCEPT     tcp  --  anywhere             anywhere             tcp dpt:https
DROP       all  --  anywhere             anywhere
```
While when performing my process the order is inverted:

```
Chain INPUT (policy ACCEPT)
target     prot opt source               destination
# My Fixed Rules
ACCEPT     all  --  anywhere             anywhere
REJECT     all  --  anywhere             loopback/8           reject-with icmp-port-unreachable
ACCEPT     all  --  anywhere             anywhere             state RELATED,ESTABLISHED
ACCEPT     tcp  --  anywhere             anywhere             tcp dpt:http
ACCEPT     tcp  --  anywhere             anywhere             tcp dpt:https
DROP       all  --  anywhere             anywhere
# F2B Rules
fail2ban-php-url-fopen  tcp  --  anywhere             anywhere             multiport dports http,https
fail2ban-apache-nohome  tcp  --  anywhere             anywhere             multiport dports http,https
```
I'm not sure if this impact how things are supposed to work.
Anyway, this process seems a bit complicated, isn't there any internal of fail2ban to simply restore the rules to iptables without restarting the whole service? It has to do at some point somehow

**Comment by mitchellkrogza** [comment]
You same bash script that does the iptables restore can just have one extra
line

sudo service fail2ban restart

And can send you a mail when done

sudo service fail2ban restart | mail -s "Iptables restored and Fail2ban
restarted" youremail@wherever.com

**Comment by sebres** [comment]
You can put together the lists before restore and do it immediately in the right order.
I mean for example something like
```bash
iptables-restore <(cat 1st.rules) <(cat 2nd.rules)
# or even:
iptables-restore <(cat 1st.rules 2nd.rules)
```
So you'll prevent an interim flushing also.

> isn't there any internal of fail2ban to simply restore the rules to iptables without restarting the whole service?

Of course we could implement something like `fail2ban-client reban` or `fail2ban-client reload --reban` to repeat banning actions for all current tickets.
But unfortunately I've currently other tasks with higher priorities, so I can't promise it will done soon, despite this feature-request looks like easy to implement .

**Comment by LeonardoGentile** [comment]
@mitchellkrogza you are correct but I found that restarting the whole service when you have several active f2b rules can take quite some time (sometimes even 15/20 seconds on a rpi). I suppose just reloading the iptables rules should take fewer time.  
Given that my original script is located in `/etc/network/if-pre-up.d`  I suppose (I hope I don't say stupid things) that just a `ifup eth0` lets you wait almost 30 sec for the interface to come up

**Comment by mitchellkrogza** [comment]
I know a restart of Fail2ban can take a good 20-30 seconds, on 0.9.x anyway
and I've heard a lot that 0.10.x is  a lot faster but have not yet migrated
my servers to 0.10.

But your iptables restore should take a mere second and Fail2ban thereafter
taking 20-30 seconds restart should not compromise you in any way. It's
like doing a simple reboot after a kernel update, server boots and loads
your default iptables rules then fail2ban starts up and takes a good few
minutes to start banning and adding entries to iptables.

Are you trying to save your iptables rules including what fail2ban has
banned and restore those all with your restore script?

On 01 Feb 2017 14:38, "LeonardoGentile" <notifications@github.com> wrote:

@mitchellkrogza <https://github.com/mitchellkrogza> you are correct but I
found that restarting the whole service when you have several active f2b
rules can take quite some time (sometimes even 15/20 seconds on a rpi). I
suppose just reloading the iptables rules should take fewer time.
Given that my original script is located in /etc/network/if-pre-up.d I
suppose (I hope I don't say stupid things) that just a ifup eth0 lets you
wait almost 30 sec for the interface to come up

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/fail2ban/fail2ban/issues/1680#issuecomment-276647433>,
or mute the thread

**Comment by sebres** [comment]
I implemented IMHO better strategy (PR #2588) as manual reban after external flushing:
automatically recognize external flush (consistency check and repair or stop/start) and reban of each single ticket that causes a failure in banned state (noticed failure from "already banned" after certain time after it got banned).
